x-pack/plugins/monitoring/server/alerts/memory_usage_alert.ts

```diff
@@ -0,0 +1,343 @@
+  AlertMessageLinkToken,
+  AlertInstanceState,
+} from './types';
+import { AlertInstance, AlertServices } from '../../../alerts/server';
+import { INDEX_PATTERN_ELASTICSEARCH, ALERT_MEMORY_USAGE } from '../../common/constants';
+import { fetchMemoryUsageNodeStats } from '../lib/alerts/fetch_memory_usage_node_stats';
+import { getCcsIndexPattern } from '../lib/alerts/get_ccs_index_pattern';
+import { AlertMessageTokenType, AlertSeverity, AlertParamType } from '../../common/enums';
+import { RawAlertInstance } from '../../../alerts/common';
+import { CommonAlertFilter, CommonAlertParams, CommonAlertParamDetail } from '../../common/types';
+import { AlertingDefaults, createLink } from './alerts_common';
+import { appendMetricbeatIndex } from '../lib/alerts/append_mb_index';
+
+interface ParamDetails {
+  [key: string]: CommonAlertParamDetail;
+}
+
+export class MemoryUsageAlert extends BaseAlert {
+  public static readonly PARAM_DETAILS: ParamDetails = {
+    threshold: {
+      label: i18n.translate('xpack.monitoring.alerts.memoryUsage.paramDetails.threshold.label', {
+        defaultMessage: `Notify when memory usage is over`,
+      }),
+      type: AlertParamType.Percentage,
+    },
+    duration: {
+      label: i18n.translate('xpack.monitoring.alerts.memoryUsage.paramDetails.duration.label', {
+        defaultMessage: `Look at the average over`,
+      }),
+      type: AlertParamType.Duration,
+    },
+  };
+  public static paramDetails = MemoryUsageAlert.PARAM_DETAILS;
+  public static readonly TYPE = ALERT_MEMORY_USAGE;
+  public static readonly LABEL = i18n.translate('xpack.monitoring.alerts.memoryUsage.label', {
+    defaultMessage: 'Memory Usage (JVM)',
+  });
+  public type = MemoryUsageAlert.TYPE;
+  public label = MemoryUsageAlert.LABEL;
+
+  protected defaultParams = {
+    threshold: 90,
+    duration: '5m',
+  };
+
+  protected actionVariables = [
+    {
+      name: 'nodes',
+      description: i18n.translate('xpack.monitoring.alerts.memoryUsage.actionVariables.nodes', {
+        defaultMessage: 'The list of nodes reporting high memory usage.',
+      }),
+    },
+    {
+      name: 'count',
+      description: i18n.translate('xpack.monitoring.alerts.memoryUsage.actionVariables.count', {
+        defaultMessage: 'The number of nodes reporting high memory usage.',
+      }),
+    },
+    ...Object.values(AlertingDefaults.ALERT_TYPE.context),
+  ];
+
+  protected async fetchData(
+    params: CommonAlertParams,
+    callCluster: any,
+    clusters: AlertCluster[],
+    uiSettings: IUiSettingsClient,
+    availableCcs: string[]
+  ): Promise<AlertData[]> {
+    let esIndexPattern = appendMetricbeatIndex(this.config, INDEX_PATTERN_ELASTICSEARCH);
+    if (availableCcs) {
+      esIndexPattern = getCcsIndexPattern(esIndexPattern, availableCcs);
+    }
+    const { duration, threshold } = params;
+    const stats = await fetchMemoryUsageNodeStats(
+      callCluster,
+      clusters,
+      esIndexPattern,
+      duration as string,
```

Comment:
The reason why I'm parsing the duration into a number (see https://github.com/elastic/kibana/blob/master/x-pack/plugins/monitoring/server/alerts/cpu_usage_alert.ts#L97) is because I don't know if we can guarantee that the format of time within the alerting framework will always be consistent with Elasticsearch date math.

Perhaps @mikecote can clarify this. Can we expect the [duration used by alerts](https://github.com/elastic/kibana/blob/master/x-pack/plugins/alerts/common/parse_duration.ts) to be directly compatible with [ES date math](https://www.elastic.co/guide/en/elasticsearch/reference/current/common-options.html#date-math)

Reply:
>Can we expect the duration used by alerts to be directly compatible with ES date math

The reason makes sense. I think it's good to side on not being directly compatible as we can't guarantee they'll always be in sync. Also that we don't support all suffixes at this time and we plan on [leveraging a library function soon](https://github.com/elastic/kibana/issues/50259).